File: Builder/lib/services/code_runner.dart
import 'dart:math' as math;

class CodeRunner {
  CodeRunner._();
  static final CodeRunner _instance = CodeRunner._();
  static CodeRunner get instance => _instance;

  /// Run code.
  /// NOTE: This is still a local simulator, not a real Python interpreter.
  Future<CodeRunResult> runCode(
    String code, {
    String language = 'python',
  }) async {
    // Simulate execution delay
    await Future.delayed(const Duration(milliseconds: 500));

    if (language.toLowerCase() != 'python') {
      return _runStringLiteralPrintOnly(code);
    }

    return _runPythonLike(code);
  }

  CodeRunResult _runStringLiteralPrintOnly(String code) {
    final printPattern = RegExp(r'print\s*\(\s*["\x27](.+?)["\x27]\s*\)');
    final matches = printPattern.allMatches(code);

    if (matches.isEmpty) {
      return const CodeRunResult(success: true, output: '(no output)');
    }

    final outputs = matches.map((m) => m.group(1) ?? '').toList();
    return CodeRunResult(success: true, output: outputs.join('\n'));
  }

  CodeRunResult _runPythonLike(String code) {
    final outputs = <String>[];
    final variables = <String, dynamic>{};

    try {
      for (final rawLine in code.split('\n')) {
        final line = _stripInlineComment(rawLine).trim();
        if (line.isEmpty) continue;

        if (_isAssignmentLine(line)) {
          final splitIndex = line.indexOf('=');
          final name = line.substring(0, splitIndex).trim();
          final expr = line.substring(splitIndex + 1).trim();
          if (!_isValidIdentifier(name)) continue;

          final value = _evalExpression(expr, variables);
          variables[name] = value;
          continue;
        }

        final printMatch = RegExp(
          r'^print\s*\(([\s\S]*)\)\s*$',
        ).firstMatch(line);
        if (printMatch != null) {
          final argText = printMatch.group(1) ?? '';
          final argExpressions = _splitTopLevel(argText, ',');
          final values = argExpressions
              .where((arg) => arg.trim().isNotEmpty)
              .map((arg) => _evalExpression(arg.trim(), variables))
              .map(_formatPythonValue)
              .toList();
          outputs.add(values.join(' '));
        }
      }
    } catch (e) {
      return CodeRunResult(
        success: false,
        output: '',
        error: 'Runtime error: $e',
      );
    }

    return CodeRunResult(
      success: true,
      output: outputs.isEmpty ? '(no output)' : outputs.join('\n'),
    );
  }

  bool _isAssignmentLine(String line) {
    if (!line.contains('=')) return false;
    if (line.contains('==') ||
        line.contains('!=') ||
        line.contains('>=') ||
        line.contains('<=') ||
        line.startsWith('print(')) {
      return false;
    }

    final splitIndex = line.indexOf('=');
    if (splitIndex <= 0) return false;
    final name = line.substring(0, splitIndex).trim();
    return _isValidIdentifier(name);
  }

  bool _isValidIdentifier(String value) {
    final idPattern = RegExp(r'^[A-Za-z_][A-Za-z0-9_]*$');
    return idPattern.hasMatch(value);
  }

  String _stripInlineComment(String line) {
    final buffer = StringBuffer();
    bool inSingle = false;
    bool inDouble = false;
    bool escaped = false;

    for (int i = 0; i < line.length; i++) {
      final ch = line[i];
      if (escaped) {
        buffer.write(ch);
        escaped = false;
        continue;
      }
      if (ch == r'\') {
        buffer.write(ch);
        escaped = true;
        continue;
      }
      if (!inDouble && ch == "'") {
        inSingle = !inSingle;
        buffer.write(ch);
        continue;
      }
      if (!inSingle && ch == '"') {
        inDouble = !inDouble;
        buffer.write(ch);
        continue;
      }
      if (!inSingle && !inDouble && ch == '#') {
        break;
      }
      buffer.write(ch);
    }

    return buffer.toString();
  }

  dynamic _evalExpression(String expr, Map<String, dynamic> variables) {
    final input = expr.trim();
    if (input.isEmpty) {
      throw const FormatException('empty expression');
    }

    if (_isWrappedByBalancedParentheses(input)) {
      return _evalExpression(input.substring(1, input.length - 1), variables);
    }

    final stringLiteral = _tryParseString(input);
    if (stringLiteral != null) return stringLiteral;

    final number = _tryParseNumber(input);
    if (number != null) return number;

    if (input == 'True') return true;
    if (input == 'False') return false;

    if (_isValidIdentifier(input) && variables.containsKey(input)) {
      return variables[input];
    }

    final fnMatch = RegExp(
      r'^([A-Za-z_][A-Za-z0-9_]*)\(([\s\S]*)\)$',
    ).firstMatch(input);
    if (fnMatch != null) {
      final fn = fnMatch.group(1)!;
      final rawArgs = fnMatch.group(2) ?? '';
      final args = _splitTopLevel(rawArgs, ',')
          .where((arg) => arg.trim().isNotEmpty)
          .map((arg) => _evalExpression(arg.trim(), variables))
          .toList();
      return _evalFunction(fn, args);
    }

    final addSub = _findTopLevelOperator(input, const ['+', '-']);
    if (addSub != null) {
      final left = _evalExpression(input.substring(0, addSub.index), variables);
      final right = _evalExpression(
        input.substring(addSub.index + addSub.operator.length),
        variables,
      );
      return _applyNumericOperator(left, right, addSub.operator);
    }

    final mulDiv = _findTopLevelOperator(input, const ['//', '*', '/', '%']);
    if (mulDiv != null) {
      final left = _evalExpression(input.substring(0, mulDiv.index), variables);
      final right = _evalExpression(
        input.substring(mulDiv.index + mulDiv.operator.length),
        variables,
      );
      return _applyNumericOperator(left, right, mulDiv.operator);
    }

    final power = _findTopLevelOperator(input, const [
      '**',
    ], rightAssociative: true);
    if (power != null) {
      final left = _evalExpression(input.substring(0, power.index), variables);
      final right = _evalExpression(
        input.substring(power.index + power.operator.length),
        variables,
      );
      return _applyNumericOperator(left, right, power.operator);
    }

    throw FormatException('unsupported expression: $input');
  }

  dynamic _evalFunction(String fn, List<dynamic> args) {
    switch (fn) {
      case 'type':
        if (args.length != 1) {
          throw FormatException(
            'type() expects 1 argument, got ${args.length}',
          );
        }
        final value = args.first;
        final typeName = switch (value) {
          int _ => 'int',
          double _ => 'float',
          bool _ => 'bool',
          String _ => 'str',
          _ => 'object',
        };
        return "<class '$typeName'>";
      case 'int':
        if (args.length != 1) {
          throw FormatException('int() expects 1 argument, got ${args.length}');
        }
        return _toNum(args.first).truncate();
      case 'float':
        if (args.length != 1) {
          throw FormatException(
            'float() expects 1 argument, got ${args.length}',
          );
        }
        return _toNum(args.first).toDouble();
      case 'round':
        if (args.length != 1) {
          throw FormatException(
            'round() expects 1 argument, got ${args.length}',
          );
        }
        return _toNum(args.first).round();
      case 'str':
        if (args.length != 1) {
          throw FormatException('str() expects 1 argument, got ${args.length}');
        }
        return _formatPythonValue(args.first);
      case 'bool':
        if (args.length != 1) {
          throw FormatException(
            'bool() expects 1 argument, got ${args.length}',
          );
        }
        final value = args.first;
        if (value is bool) return value;
        if (value is num) return value != 0;
        if (value is String) return value.isNotEmpty;
        return value != null;
      default:
        throw FormatException('unsupported function: $fn');
    }
  }

  dynamic _applyNumericOperator(dynamic left, dynamic right, String op) {
    final a = _toNum(left);
    final b = _toNum(right);

    switch (op) {
      case '+':
        return a + b;
      case '-':
        return a - b;
      case '*':
        return a * b;
      case '/':
        return a / b;
      case '//':
        return (a / b).floor();
      case '%':
        return a % b;
      case '**':
        return _powNum(a, b);
      default:
        throw FormatException('unsupported operator: $op');
    }
  }

  num _powNum(num base, num exponent) {
    if (exponent is int) {
      var result = 1.0;
      for (int i = 0; i < exponent.abs(); i++) {
        result *= base;
      }
      return exponent < 0 ? 1 / result : result;
    }
    // Fallback for non-integer exponents.
    return math.pow(base.toDouble(), exponent.toDouble());
  }

  num _toNum(dynamic value) {
    if (value is num) return value;
    if (value is String) {
      final parsed = num.tryParse(value.trim());
      if (parsed != null) return parsed;
    }
    throw FormatException('expected numeric value, got $value');
  }

  String _formatPythonValue(dynamic value) {
    if (value == null) return 'None';
    if (value is bool) return value ? 'True' : 'False';
    if (value is double) {
      if (value == value.truncateToDouble()) {
        return value.toStringAsFixed(1);
      }
      return value.toString();
    }
    return value.toString();
  }

  String? _tryParseString(String input) {
    if (input.length < 2) return null;
    final first = input[0];
    final last = input[input.length - 1];
    if (!((first == "'" && last == "'") || (first == '"' && last == '"'))) {
      return null;
    }
    return input.substring(1, input.length - 1);
  }

  num? _tryParseNumber(String input) {
    final intValue = int.tryParse(input);
    if (intValue != null) return intValue;
    return double.tryParse(input);
  }

  bool _isWrappedByBalancedParentheses(String input) {
    if (!input.startsWith('(') || !input.endsWith(')')) return false;

    int depth = 0;
    bool inSingle = false;
    bool inDouble = false;
    bool escaped = false;

    for (int i = 0; i < input.length; i++) {
      final ch = input[i];
      if (escaped) {
        escaped = false;
        continue;
      }
      if (ch == r'\') {
        escaped = true;
        continue;
      }
      if (!inDouble && ch == "'") {
        inSingle = !inSingle;
        continue;
      }
      if (!inSingle && ch == '"') {
        inDouble = !inDouble;
        continue;
      }
      if (inSingle || inDouble) continue;

      if (ch == '(') depth += 1;
      if (ch == ')') depth -= 1;
      if (depth == 0 && i != input.length - 1) return false;
    }
    return depth == 0;
  }

  List<String> _splitTopLevel(String input, String delimiter) {
    if (input.trim().isEmpty) return const [];
    final parts = <String>[];
    int start = 0;
    int depth = 0;
    bool inSingle = false;
    bool inDouble = false;
    bool escaped = false;

    for (int i = 0; i < input.length; i++) {
      final ch = input[i];
      if (escaped) {
        escaped = false;
        continue;
      }
      if (ch == r'\') {
        escaped = true;
        continue;
      }
      if (!inDouble && ch == "'") {
        inSingle = !inSingle;
        continue;
      }
      if (!inSingle && ch == '"') {
        inDouble = !inDouble;
        continue;
      }
      if (inSingle || inDouble) continue;

      if (ch == '(') {
        depth += 1;
      } else if (ch == ')') {
        depth -= 1;
      } else if (depth == 0 && input.substring(i).startsWith(delimiter)) {
        parts.add(input.substring(start, i).trim());
        start = i + delimiter.length;
        i += delimiter.length - 1;
      }
    }

    parts.add(input.substring(start).trim());
    return parts;
  }

  _TopLevelOperator? _findTopLevelOperator(
    String input,
    List<String> operators, {
    bool rightAssociative = false,
  }) {
    final sortedOperators = [...operators]
      ..sort((a, b) => b.length.compareTo(a.length));
    final candidates = <_TopLevelOperator>[];

    int depth = 0;
    bool inSingle = false;
    bool inDouble = false;
    bool escaped = false;

    for (int i = 0; i < input.length; i++) {
      final ch = input[i];
      if (escaped) {
        escaped = false;
        continue;
      }
      if (ch == r'\') {
        escaped = true;
        continue;
      }
      if (!inDouble && ch == "'") {
        inSingle = !inSingle;
        continue;
      }
      if (!inSingle && ch == '"') {
        inDouble = !inDouble;
        continue;
      }
      if (inSingle || inDouble) continue;

      if (ch == '(') {
        depth += 1;
        continue;
      }
      if (ch == ')') {
        depth -= 1;
        continue;
      }
      if (depth != 0) continue;

      for (final op in sortedOperators) {
        if (i + op.length > input.length) continue;
        if (input.substring(i, i + op.length) != op) continue;

        if (op == '*' &&
            ((i > 0 && input[i - 1] == '*') ||
                (i + 1 < input.length && input[i + 1] == '*'))) {
          continue;
        }
        if (op == '/' &&
            ((i > 0 && input[i - 1] == '/') ||
                (i + 1 < input.length && input[i + 1] == '/'))) {
          continue;
        }
        if ((op == '+' || op == '-') && _isUnaryOperator(input, i)) {
          continue;
        }
        if (i == 0 || i + op.length >= input.length) continue;
        candidates.add(_TopLevelOperator(index: i, operator: op));
        i += op.length - 1;
        break;
      }
    }

    if (candidates.isEmpty) return null;
    return rightAssociative ? candidates.first : candidates.last;
  }

  bool _isUnaryOperator(String input, int index) {
    if (index == 0) return true;
    int i = index - 1;
    while (i >= 0 && input[i].trim().isEmpty) {
      i -= 1;
    }
    if (i < 0) return true;
    final prev = input[i];
    return prev == '(' ||
        prev == ',' ||
        prev == '+' ||
        prev == '-' ||
        prev == '*' ||
        prev == '/' ||
        prev == '%' ||
        prev == '=';
  }
}

class _TopLevelOperator {
  final int index;
  final String operator;

  const _TopLevelOperator({required this.index, required this.operator});
}

/// Code execution result
class CodeRunResult {
  final bool success;
  final String output;
  final String? error;

  const CodeRunResult({
    required this.success,
    required this.output,
    this.error,
  });
}

File: packages/viewer-react/src/features/editor/codeRunner.ts
type RunnerSuccess = {
  success: true;
  output: string;
};

type RunnerFailure = {
  success: false;
  error: string;
};

export type CodeRunResult = RunnerSuccess | RunnerFailure;

type OperatorMatch = {
  index: number;
  operator: string;
};

export async function runCode(
  code: string,
  {
    language = 'python',
    signal,
  }: {
    language?: string;
    signal?: AbortSignal;
  } = {},
): Promise<CodeRunResult> {
  await delay(450, signal);

  if (signal?.aborted) {
    throw new DOMException('Code execution aborted.', 'AbortError');
  }

  if (language.toLowerCase() !== 'python') {
    return runStringLiteralPrintOnly(code);
  }

  return runPythonLike(code);
}

function delay(ms: number, signal?: AbortSignal) {
  return new Promise<void>((resolve, reject) => {
    const timeout = window.setTimeout(() => {
      cleanup();
      resolve();
    }, ms);

    function onAbort() {
      cleanup();
      reject(new DOMException('Code execution aborted.', 'AbortError'));
    }

    function cleanup() {
      window.clearTimeout(timeout);
      signal?.removeEventListener('abort', onAbort);
    }

    signal?.addEventListener('abort', onAbort, { once: true });
  });
}

function runStringLiteralPrintOnly(code: string): CodeRunResult {
  const printPattern = /print\s*\(\s*["'](.+?)["']\s*\)/g;
  const outputs = Array.from(code.matchAll(printPattern)).map((match) => match[1] ?? '');

  return {
    success: true,
    output: outputs.length > 0 ? outputs.join('\n') : '(no output)',
  };
}

function runPythonLike(code: string): CodeRunResult {
  const outputs: string[] = [];
  const variables = new Map<string, unknown>();

  try {
    for (const rawLine of code.split('\n')) {
      const line = stripInlineComment(rawLine).trim();
      if (!line) {
        continue;
      }

      if (isAssignmentLine(line)) {
        const splitIndex = line.indexOf('=');
        const identifier = line.slice(0, splitIndex).trim();
        const expression = line.slice(splitIndex + 1).trim();

        if (isValidIdentifier(identifier)) {
          variables.set(identifier, evaluateExpression(expression, variables));
        }

        continue;
      }

      const printMatch = /^print\s*\(([\s\S]*)\)\s*$/.exec(line);
      if (printMatch) {
        const expressions = splitTopLevel(printMatch[1] ?? '', ',')
          .map((part) => part.trim())
          .filter(Boolean);
        const values = expressions.map((expression) =>
          formatPythonValue(evaluateExpression(expression, variables)),
        );
        outputs.push(values.join(' '));
      }
    }
  } catch (error) {
    return {
      success: false,
      error: `Runtime error: ${error instanceof Error ? error.message : String(error)}`,
    };
  }

  return {
    success: true,
    output: outputs.length > 0 ? outputs.join('\n') : '(no output)',
  };
}

function stripInlineComment(line: string) {
  let inSingle = false;
  let inDouble = false;
  let escaped = false;
  let result = '';

  for (const character of line) {
    if (escaped) {
      result += character;
      escaped = false;
      continue;
    }

    if (character === '\\') {
      result += character;
      escaped = true;
      continue;
    }

    if (!inDouble && character === "'") {
      inSingle = !inSingle;
      result += character;
      continue;
    }

    if (!inSingle && character === '"') {
      inDouble = !inDouble;
      result += character;
      continue;
    }

    if (!inSingle && !inDouble && character === '#') {
      break;
    }

    result += character;
  }

  return result;
}

function isAssignmentLine(line: string) {
  if (!line.includes('=')) {
    return false;
  }

  if (
    line.includes('==') ||
    line.includes('!=') ||
    line.includes('>=') ||
    line.includes('<=') ||
    line.startsWith('print(')
  ) {
    return false;
  }

  const splitIndex = line.indexOf('=');
  return splitIndex > 0 && isValidIdentifier(line.slice(0, splitIndex).trim());
}

function isValidIdentifier(value: string) {
  return /^[A-Za-z_][A-Za-z0-9_]*$/.test(value);
}

function evaluateExpression(expression: string, variables: Map<string, unknown>): unknown {
  const input = expression.trim();
  if (!input) {
    throw new Error('empty expression');
  }

  if (isWrappedByBalancedParentheses(input)) {
    return evaluateExpression(input.slice(1, -1), variables);
  }

  const stringLiteral = tryParseString(input);
  if (stringLiteral !== null) {
    return stringLiteral;
  }

  const numberLiteral = tryParseNumber(input);
  if (numberLiteral !== null) {
    return numberLiteral;
  }

  if (input === 'True') {
    return true;
  }

  if (input === 'False') {
    return false;
  }

  if (isValidIdentifier(input) && variables.has(input)) {
    return variables.get(input);
  }

  const functionMatch = /^([A-Za-z_][A-Za-z0-9_]*)\(([\s\S]*)\)$/.exec(input);
  if (functionMatch) {
    const functionName = functionMatch[1] ?? '';
    const rawArgs = functionMatch[2] ?? '';
    const args = splitTopLevel(rawArgs, ',')
      .map((part) => part.trim())
      .filter(Boolean)
      .map((part) => evaluateExpression(part, variables));
    return evaluateFunction(functionName, args);
  }

  const addSubMatch = findTopLevelOperator(input, ['+', '-']);
  if (addSubMatch) {
    const left = evaluateExpression(input.slice(0, addSubMatch.index), variables);
    const right = evaluateExpression(
      input.slice(addSubMatch.index + addSubMatch.operator.length),
      variables,
    );
    return applyNumericOperator(left, right, addSubMatch.operator);
  }

  const mulDivMatch = findTopLevelOperator(input, ['//', '*', '/', '%']);
  if (mulDivMatch) {
    const left = evaluateExpression(input.slice(0, mulDivMatch.index), variables);
    const right = evaluateExpression(
      input.slice(mulDivMatch.index + mulDivMatch.operator.length),
      variables,
    );
    return applyNumericOperator(left, right, mulDivMatch.operator);
  }

  const powerMatch = findTopLevelOperator(input, ['**'], true);
  if (powerMatch) {
    const left = evaluateExpression(input.slice(0, powerMatch.index), variables);
    const right = evaluateExpression(
      input.slice(powerMatch.index + powerMatch.operator.length),
      variables,
    );
    return applyNumericOperator(left, right, powerMatch.operator);
  }

  throw new Error(`unsupported expression: ${input}`);
}

function evaluateFunction(functionName: string, args: unknown[]) {
  switch (functionName) {
    case 'type': {
      if (args.length !== 1) {
        throw new Error(`type() expects 1 argument, got ${args.length}`);
      }

      const value = args[0];
      const typeName =
        typeof value === 'boolean'
          ? 'bool'
          : typeof value === 'number'
          ? Number.isInteger(value)
            ? 'int'
            : 'float'
          : typeof value === 'string'
          ? 'str'
          : 'object';
      return `<class '${typeName}'>`;
    }
    case 'int':
      return Math.trunc(toNumber(args, functionName));
    case 'float':
      return toNumber(args, functionName);
    case 'round':
      return Math.round(toNumber(args, functionName));
    case 'str':
      return formatPythonValue(expectSingleArg(args, functionName));
    case 'bool': {
      const value = expectSingleArg(args, functionName);
      if (typeof value === 'boolean') {
        return value;
      }
      if (typeof value === 'number') {
        return value !== 0;
      }
      if (typeof value === 'string') {
        return value.length > 0;
      }
      return Boolean(value);
    }
    default:
      throw new Error(`unsupported function: ${functionName}`);
  }
}

function expectSingleArg(args: unknown[], functionName: string) {
  if (args.length !== 1) {
    throw new Error(`${functionName}() expects 1 argument, got ${args.length}`);
  }

  return args[0];
}

function toNumber(args: unknown[], functionName: string) {
  const value = expectSingleArg(args, functionName);
  const numericValue =
    typeof value === 'number'
      ? value
      : typeof value === 'string'
      ? Number(value)
      : NaN;

  if (Number.isNaN(numericValue)) {
    throw new Error(`${functionName}() could not convert ${String(value)}`);
  }

  return numericValue;
}

function applyNumericOperator(left: unknown, right: unknown, operator: string) {
  if (operator === '+' && (typeof left === 'string' || typeof right === 'string')) {
    return `${formatPythonValue(left)}${formatPythonValue(right)}`;
  }

  const leftNumber = typeof left === 'number' ? left : Number(left);
  const rightNumber = typeof right === 'number' ? right : Number(right);

  if (Number.isNaN(leftNumber) || Number.isNaN(rightNumber)) {
    throw new Error(`unsupported operands for ${operator}: ${String(left)} and ${String(right)}`);
  }

  switch (operator) {
    case '+':
      return leftNumber + rightNumber;
    case '-':
      return leftNumber - rightNumber;
    case '*':
      return leftNumber * rightNumber;
    case '/':
      return leftNumber / rightNumber;
    case '//':
      return Math.floor(leftNumber / rightNumber);
    case '%':
      return leftNumber % rightNumber;
    case '**':
      return leftNumber ** rightNumber;
    default:
      throw new Error(`unsupported operator: ${operator}`);
  }
}

function formatPythonValue(value: unknown) {
  if (typeof value === 'boolean') {
    return value ? 'True' : 'False';
  }

  return String(value);
}

function tryParseString(value: string) {
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    return value.slice(1, -1);
  }

  return null;
}

function tryParseNumber(value: string) {
  if (!/^[-+]?\d+(\.\d+)?$/.test(value)) {
    return null;
  }

  return value.includes('.') ? Number.parseFloat(value) : Number.parseInt(value, 10);
}

function isWrappedByBalancedParentheses(value: string) {
  if (!value.startsWith('(') || !value.endsWith(')')) {
    return false;
  }

  let depth = 0;
  let inSingle = false;
  let inDouble = false;
  let escaped = false;

  for (let index = 0; index < value.length; index += 1) {
    const character = value[index]!;

    if (escaped) {
      escaped = false;
      continue;
    }

    if (character === '\\') {
      escaped = true;
      continue;
    }

    if (!inDouble && character === "'") {
      inSingle = !inSingle;
      continue;
    }

    if (!inSingle && character === '"') {
      inDouble = !inDouble;
      continue;
    }

    if (inSingle || inDouble) {
      continue;
    }

    if (character === '(') {
      depth += 1;
    } else if (character === ')') {
      depth -= 1;
      if (depth === 0 && index !== value.length - 1) {
        return false;
      }
    }
  }

  return depth === 0;
}

function findTopLevelOperator(
  value: string,
  operators: string[],
  rightAssociative = false,
): OperatorMatch | null {
  let depth = 0;
  let inSingle = false;
  let inDouble = false;
  let escaped = false;
  const indexes = [...Array(value.length).keys()];
  const sequence = rightAssociative ? indexes.reverse() : indexes;

  for (const index of sequence) {
    const character = value[index]!;

    if (escaped) {
      escaped = false;
      continue;
    }

    if (character === '\\') {
      escaped = true;
      continue;
    }

    if (!inDouble && character === "'") {
      inSingle = !inSingle;
      continue;
    }

    if (!inSingle && character === '"') {
      inDouble = !inDouble;
      continue;
    }

    if (inSingle || inDouble) {
      continue;
    }

    if (character === ')') {
      depth += 1;
      continue;
    }

    if (character === '(') {
      depth -= 1;
      continue;
    }

    if (depth !== 0) {
      continue;
    }

    for (const operator of operators) {
      const segment = rightAssociative
        ? value.slice(index - operator.length + 1, index + 1)
        : value.slice(index, index + operator.length);

      if (segment !== operator) {
        continue;
      }

      const startIndex = rightAssociative ? index - operator.length + 1 : index;
      const previous = value[startIndex - 1] ?? '';

      if (
        (operator === '+' || operator === '-') &&
        (startIndex === 0 || /[,(+\-*/%]/.test(previous))
      ) {
        continue;
      }

      return { index: startIndex, operator };
    }
  }

  return null;
}

function splitTopLevel(value: string, separator: string) {
  const parts: string[] = [];
  let depth = 0;
  let inSingle = false;
  let inDouble = false;
  let escaped = false;
  let current = '';

  for (const character of value) {
    if (escaped) {
      current += character;
      escaped = false;
      continue;
    }

    if (character === '\\') {
      current += character;
      escaped = true;
      continue;
    }

    if (!inDouble && character === "'") {
      inSingle = !inSingle;
      current += character;
      continue;
    }

    if (!inSingle && character === '"') {
      inDouble = !inDouble;
      current += character;
      continue;
    }

    if (inSingle || inDouble) {
      current += character;
      continue;
    }

    if (character === '(') {
      depth += 1;
      current += character;
      continue;
    }

    if (character === ')') {
      depth -= 1;
      current += character;
      continue;
    }

    if (character === separator && depth === 0) {
      parts.push(current);
      current = '';
      continue;
    }

    current += character;
  }

  parts.push(current);
  return parts;
}
